packages/controls/src/widget_dom_events.ts

```diff
@@ -0,0 +1,276 @@
+import {
+    WidgetModel, WidgetView, DOMWidgetView, unpack_models
+} from '@jupyter-widgets/base';
+
+import * as _ from 'underscore';
+
+// The names in the lists below are what will be sent as part of the
+// event message to the backend.
+// The actual list is constructed in _send_dom_message
+
+let common_event_message_names = [
+    'altKey',
+    'ctrlKey',
+    'metaKey',
+    'shiftKey',
+    'type',
+    'timeStamp'
+]
+
+let mouse_standard_event_message_names = [
+    'button',
+    'buttons',
+    'clientX',
+    'clientY',
+    'layerX',
+    'layerY',
+    'movementX',
+    'movementY',
+    'offsetX',
+    'offsetY',
+    'pageX',
+    'pageY',
+    'screenX',
+    'screenY',
+    'x',
+    'y'
+]
+
+let mouse_added_event_message_names = [
+    'arrayX',
+    'arrayY',
+    'relativeX',
+    'relativeY'
+]
+
+let wheel_standard_event_names = [
+    'deltaX',
+    'deltaY',
+    'deltaZ',
+    'deltaMode'
+]
+
+let drag_standard_event_names = [
+    'dataTransfer'
+]
+
+let key_standard_event_names = [
+    'code',
+    'key',
+    'location',
+    'repeat'
+]
+
+function _get_position(view, event) {
+    // Return something like the position relative to the element to which
+    // the listener is attached. This is essentially what layerX and layerY
+    // are supposed to be (and are in chrome) but those event properties have
+    // red box warnings in the MDN documentation that they are not part of any
+    // standard and are not on any standards tracks, so get what we need here.
+    var bounding_rect = view.el.getBoundingClientRect();
+    var y_offset = bounding_rect.top;
+    var x_offset = bounding_rect.left;
+    return {
+        'x': Math.round(event.clientX - x_offset),
+        'y': Math.round(event.clientY - y_offset)
```

Comment:
What is the reason for the rounding?

Reply:
The bounding rectangles return fractional pixels but it doesn't seem correct to pass that directly on to the user. To me it implies sub-pixel accuracy that isn't really there.

Reply:
Ah, but it might be correct if the user has scaled the widget, and a pixel in the image is taking more than one pixel in the display.  At least it would be nice to get the floating point values.